docs: add customization guide

marquiz opened the pull request:
Add a separate customization guide. Move documentation of the custom and
local sources there. Also, cover the new NodeFeatureRules custom
resource and the new expression-based label rule format.

This patch also simplifies the "Feature labels" page, describing
built-in labels. Reformat the tables describing feature labels.

marquiz commented:
/assign @zvonkok  @ArangoGutierrez 
ping @mythi 

It's possibly easier to review by looking at the end result. This can be seen at gh-pages of my personal repo:
https://marquiz.github.io/node-feature-discovery/master/get-started/features.html
https://marquiz.github.io/node-feature-discovery/master/advanced/customization-guide.html

Some notes/questions:
- I simplified the page [describing built-in labels](https://marquiz.github.io/node-feature-discovery/master/get-started/features.html) quite a bit, also changed the formatting of the tables. WDYT?
- I tried to make the guide more example-driven, starting with a simple example and describing the details after that.
- Terminology is always hard, e.g. [feature types](https://marquiz.github.io/node-feature-discovery/master/advanced/customization-guide.html#feature-types)
- Customization guide is quite long page, could/should it be split into multiple sub-pages 🤔

zvonkok commented:
Looking good to me, but I would like to have someone else to take a look, just to be sure we're not missing anything. @ArangoGutierrez PTAL, thanks!

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1007 @@ published: false
+  rules:
+    - name: "my sgx feature rule"
+      labels:
+        "my-sgx-feature": "true"
```
nit: even if this is just an example, to make it work in all cases add (or change `cpu.cpuid`):
```
     - feature: cpu.sgx
        matchExpressions:
          enabled: {op: IsTrue}
```

marquiz replied:
Yeah, I know. The idea behind the example was to have a rule with more than one matcher. Maybe I'll change this to something completely different like `cpu.pststate.turbo` and `kernel.config.PREEMPT` or smth 🤔

marquiz replied:
Changed this to
```
      matchFeatures:
        - feature: kernel.loadedmodule
          matchExpressions:
            dummy: {op: Exists}
        - feature: kernel.config
          matchExpressions:
            X86: {op: In, value: ["y"]}
```

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1007 @@ published: false
+
+- the CPU has SGX capability
+- X86_SGX option has been enabled in the kernel
+
```
maybe add the basic `$ kubectl apply -f nfd-rule-sgx.yaml` to show how these are deployed, or maybe also the syntax how CRs can be added via `kustomization.yaml`.

marquiz replied:
Makes sense. Maybe we could even include the sample CR in the repo 🧐

marquiz replied:
Added a kubectl example and added a sample CR yaml in the repo

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1007 @@ published: false
+overriding built-in labels.
+
+The *local* feature source has two methods for detecting features, hooks and
+feature files.
```
some readers might ask can `local` sources info be used with `matchFeatures` with the two other customization features?

marquiz replied:
Good point. Any suggestion how to fix this? *The features discovered by the local source can further be used in label rules of NodeFeatureRule objects and the custom source*(?)

mythi replied:
The proposal looks good to me!

marquiz replied:
Changed along these lines

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1007 @@ published: false
+
+One use case for the hooks and/or feature files is detecting features in other
+Pods outside NFD, e.g. in Kubernetes device plugins. By using the same
+`hostPath` mounts for `/etc/kubernetes/node-feature-discovery/source.d/` and
```
it's not NFD specific as such but on SELinux enabled platforms it's more than just `hostPath`. Maybe add a note here that it needs to be taken into account too or, preferably, the suggestion what needs to be done to overcome that limitation.

marquiz replied:
TBH, I haven't tried this in an SELinux-enabled environment. Any suggestion what to add? *Note that if SELinux is enabled you need to correctly set `seLinuxOptions` in the security context of all pods mounting the same host path.*(?)

mythi replied:
I was thinking @ArangoGutierrez might be able to help with that.

ArangoGutierrez replied:
OpenShift has SELinux enabled, or you can enable it on MiniKube for testing.
Whit that said, yes, I usually get around that by running the 2 containers in the same pod, that way they share namespace and you can easily share mounts. (This for documentation comes handy we don't want people to think they can run pod's on other nodes or something, it has to be same node, and preferably same pod)

mythi replied:
This doc talks about pods outside nfd. My comment was should we add note that they need to run with some specific selinux label to be able to write to nfd dirs?

mythi commented on docs/advanced/customization-guide.md:
````diff
@@ -16,8 +15,1007 @@ published: false
+```
+
+It specifies one rule which creates node label
+`feature.node.kubenernetes.io/my-sgx-feature=true` if both of the following
````
`custom` rule adds it's own prefix:
`feature.node.kubenernetes.io/custom-my-sgx-feature=true`

marquiz replied:
It shouldn't anymore, after #697.

The legacy `matchOn` rules add the prefix, though. Maybe we should add a note about this 🤔

mythi replied:
Good to know!

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1007 @@ published: false
+      labels:
+        "my-sgx-feature": "true"
+      matchFeatures:
+        - feature: cpu.cpuid
```
the earlier comment about `cpu.sgx` applies here too.

marquiz replied:
Changed the example ximilarly to the CR example, using `kernel.loadedmodule.dummy`

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1007 @@ published: false
+The namespace part (i.e. prefix) of the labels is controlled by nfd:
+
+- All built-in labels use `feature.node.kubernetes.io`. This is also
+  the default for user defined features that don't specify any namespace.
```
and that `<name>` includes the source as prefix?

marquiz replied:
For built-in labels, yes. But I don't think it needs to be stated here(?)

marquiz replied:
Did not change this

mythi commented on docs/get-started/features.md:
```diff
@@ -137,519 +105,111 @@ See the full list in [github.com/klauspost/cpuid][klauspost-cpuid].
-|         | configured | SR-IOV virtual functions have been configured
+| Feature     | Value | Description
+| ----------- | ----- | -----------
+| **`network-sriov.capable`**    | true | [Single Root Input/Output Virtualization][sriov] (SR-IOV) enabled Network Interface Card(s) present
```
did you add some new labels for `network`?

marquiz replied:
Nope, this has been there for a while, already

mythi replied:
I was referring to #660 but after taking a closer look it did not change the existing labels.

mythi reviewed (commented):
A couple of comments/questions. (I skipped the match syntax part)

marquiz commented:
Thanks for the feedback @mythi! Addressed most of the feedback, hostpath mounts with selinux still open.

Review changes are now in a separate commit. Will squash after we're happy
/hold

mythi commented on docs/advanced/customization-guide.md:
```diff
@@ -16,8 +15,1030 @@ published: false
+
+### NodeFeatureRule controller
+
+NFD-Master acts as the controller for NodeFeatureRule objects. It applies these
```
```suggestion
NFD-Master acts as the controller for `NodeFeatureRule` objects. It applies these
```

ditto for other cases too

marquiz replied:
Yeah, maybe it makes it more readable. Changed as suggested. I also changed *local* and *custom* similarly

mythi reviewed (commented):
LGTM, one small suggestion

marquiz commented:
@mythi I addressed your comment in a separate patch. You check if you want.

@ArangoGutierrez anything to add? If not, I'll squash this and we could possibly merge

ArangoGutierrez commented:
> @mythi I addressed your comment in a separate patch. You check if you want.
> 
> @ArangoGutierrez anything to add? If not, I'll squash this and we could possibly merge

Squash it!

marquiz commented:
Rebased and squashed

mythi reviewed (approved):
+1

marquiz commented:
/unhold

marquiz commented:
ping @zvonkok @ArangoGutierrez

ArangoGutierrez reviewed (approved):
/lgtm
